subprojects/core/src/test/groovy/org/gradle/api/internal/tasks/properties/DefaultPropertyWalkerTest.groovy

```diff
@@ -111,15 +109,55 @@ class DefaultPropertyWalkerTest extends AbstractProjectBuilderSpec {
         1 * visitor.visitInputProperty({ it.propertyName == 'bean' && it.value == null })
     }
 
-    @Unroll
-    def "correct implementation for #type coerced to Action is tracked"() {
-        expect:
-        DefaultPropertyWalker.getImplementationClass(implementation as Action) == implementation.getClass()
+    def "cycle in nested inputs is detected"() {
+        def task = project.tasks.create("myTask", TaskWithNestedObject)
+        def cycle = new Tree(value: "cycle", left: new Tree(value: "left"))
+        cycle.right = cycle
+        task.nested = new Tree(value: "first", left: new Tree(value: "left"), right: new Tree(value: "deeper", left: cycle, right: new Tree(value: "no-cycle")))
 
-        where:
-        type      | implementation
-        "Closure" | { it }
-        "Action"  |  new Action<String>() { @Override void execute(String s) {} }
+        when:
+        visitProperties(task)
+
+        then:
+        IllegalStateException e = thrown(IllegalStateException)
+        e.message == "Cycles between nested beans are not allowed. Cycle detected between: 'nested.right.left' and 'nested.right.left.right'."
+    }
+
+    def "cycle in nested input and task itself is detected"() {
+        def task = project.tasks.create("myTask", TaskWithNestedObject)
+        task.nested = new Tree(value: "root", left: task, right: new Tree(value: "right"))
+
+        when:
+        visitProperties(task)
+
+        then:
+        IllegalStateException e = thrown(IllegalStateException)
+        e.message == "Cycles between nested beans are not allowed. Cycle detected between: '<root>' and 'nested.left'."
+    }
```

Comment:
Needs test where nested bean `equals()` parent bean, but there's no cycle.

Reply:
Added the test.